# Regression Week 2: Multiple Regression (gradient descent)

In the first notebook we explored multiple regression using Turi Create. Now we will use Turi Create along with numpy to solve for the regression weights with gradient descent.

In this notebook we will cover estimating multiple regression weights via gradient descent. We will:
* Add a constant column of 1's to a Turi Create SFrame to account for the intercept
* Convert an SFrame into a Numpy array
* Write a `predict_output()` function using Numpy
* Write a numpy function to compute the derivative of the regression weights with respect to a single feature
* Write gradient descent function to compute the regression weights given an initial weight vector, step size and tolerance.
* Use the gradient descent function to estimate regression weights for multiple features

# Fire up Turi Create

Make sure you have the latest version of Turi Create

In [1]:
import turicreate

# Load in house sales data

Dataset is from house sales in King County, the region where the city of Seattle, WA is located.

In [2]:
sales = turicreate.SFrame('./data/home_data.sframe/')

If we want to do any "feature engineering" like creating new features or adjusting existing ones we should do this directly using the SFrames as seen in the other Week 2 notebook. For this notebook, however, we will work with the existing features.

# Convert to Numpy Array

Although SFrames offer a number of benefits to users (especially when using Big Data and built-in Turi Create functions) in order to understand the details of the implementation of algorithms it's important to work with a library that allows for direct (and optimized) matrix operations. Numpy is a Python solution to work with matrices (or any multi-dimensional "array").

Recall that the predicted value given the weights and the features is just the dot product between the feature and weight vector. Similarly, if we put all of the features row-by-row in a matrix then the predicted value for *all* the observations can be computed by right multiplying the "feature matrix" by the "weight vector". 

First we need to take the SFrame of our data and convert it into a 2D numpy array (also called a matrix). To do this we use Turi Create's built in .to_dataframe() which converts the SFrame into a Pandas (another python library) dataframe. We can then use Panda's .as_matrix() to convert the dataframe into a numpy matrix.

In [3]:
import numpy as np # note this allows us to refer to numpy as np instead 

Now we will write a function that will accept an SFrame, a list of feature names (e.g. ['sqft_living', 'bedrooms']) and an target feature e.g. ('price') and will return two things:
* A numpy matrix whose columns are the desired features plus a constant column (this is how we create an 'intercept')
* A numpy array containing the values of the output

With this in mind, complete the following function:

In [4]:
def get_numpy_data(data_sf, features, output):
    data_sf['constant'] = 1.0                # this is how you add a constant column to an SFrame
    features = ['constant', *features]       # add the column 'constant' to the front of the features list so that we can extract it...
    features_sf = data_sf[features]          # select the columns of data SFrame given by the features list (including constant):  
    feature_matrix = features_sf.to_numpy()  # convert the features_SFrame into a numpy matrix:
    output_sarray = data_sf[output]          # assign the column of data_sf associated with the output to the SArray output_sarray
    output_array = output_sarray.to_numpy()  # convert the SArray into a numpy array by first converting it to a list
    return (feature_matrix, output_array)

For testing let's use the 'sqft_living' feature and a constant as our features and price as our output:

In [5]:
(example_features, example_output) = get_numpy_data(sales, ['sqft_living'], 'price') # the [] around 'sqft_living' makes it a list
print(example_features[0,:])  # this accesses the first row of the data the ':' indicates 'all columns'
print(example_output[0])      # and the corresponding output

[1.00e+00 1.18e+03]
221900.0


# Predicting output given regression weights

Suppose we had the weights `[1.0, 1.0]` and the features `[1.0, 1180.0]` and we wanted to compute the predicted output `1.0 * 1.0 + 1.0 * 1180.0 = 1181.0` this is the dot product between these two arrays. If they're numpy arrays we can use `np.dot()` to compute this:

In [6]:
my_weights = np.array([1., 1.]) # the example weights
my_features = example_features[0,] # we'll use the first data point
predicted_value = np.dot(my_features, my_weights)
predicted_value

1181.0

np.dot() also works when dealing with a matrix and a vector. Recall that the predictions from all the observations is just the *right* (as in weights on the right) dot product between the features *matrix* and the weights *vector*.  

With this in mind finish the following `predict_output` function to compute the predictions for an entire matrix of features given the matrix and the weights:

In [7]:
def predict_output(feature_matrix, weights):
    # assume feature_matrix is a numpy matrix containing the features as columns and weights is a corresponding numpy array
    return np.dot(feature_matrix, weights)

If you want to test your code run the following cell:

In [8]:
test_predictions = predict_output(example_features, my_weights)
assert test_predictions[0] == 1181.0  # should be 1181.0
assert test_predictions[1] == 2571.0  # should be 2571.0

# expect no output - good

# Computing the Derivative

We are now going to move to computing the derivative of the regression cost function. Recall that the cost function is the sum over the data points of the squared difference between an observed output and a predicted output.

Since the derivative of a sum is the sum of the derivatives we can compute the derivative for a single data point and then sum over data points. We can write the squared difference between the observed output and predicted output for a single point as follows:

$$(w[0]\times[CONSTANT] + w[1]\times[feature_1] + ... + w[i]\times[feature_i] + ... +  w[k]\times[feature_k] - output)^2$$

Where we have $k$ features and a constant. So the derivative with respect to weight $w[i]$ by the chain rule is:

$$2\times(w[0]\times[CONSTANT] + w[1]\times[feature_1] + ... + w[i]\times[feature_i] + ... +  w[k]\times[feature_k] - output)\times[feature_i]$$

The term inside the parenthesis is just the error (difference between prediction and output). So we can re-write this as:

$$2\times error\times[feature_i]$$

That is, the derivative for the weight for feature $i$ is:
  - the sum (over data points) of 2 times the product of the error and the feature itself.   
    In the case of the constant then this is just twice the sum of the errors!


Recall that twice the sum of the product of two vectors is just:
  - twice the dot product of the two vectors.  
    Therefore the derivative for the weight for $feature_i$ is just two times the dot product between the values of $feature_i$ and the current errors. 

With this in mind complete the following derivative function which computes the derivative of the weight given the value of the feature (over all data points) and the errors (over all data points).

In [9]:
def feature_derivative(errors, feature):
    # Assume that errors and feature are both numpy arrays of the same length (number of data points)
    return 2 * np.dot(feature, errors) # compute twice the dot product of these vectors as 'derivative' and return the value

To test your feature derivartive run the following:

In [10]:
(example_features, example_output) = get_numpy_data(sales, ['sqft_living'], 'price') 
my_weights = np.array([0., 0.]) # this makes all the predictions 0
test_predictions = predict_output(example_features, my_weights) 

# just like SFrames 2 numpy arrays can be elementwise subtracted with '-': 
errors = test_predictions - example_output # prediction errors in this case is just the -example_output
feature = example_features[:, 0]           # let's compute the derivative with respect to 'constant', the ":" indicates "all rows"
derivative = feature_derivative(errors, feature)
alt_derivative = -np.sum (example_output) * 2

print(f"derivative: {derivative:2.5e} / alt_derivative: {alt_derivative:2.5e}")
assert derivative == alt_derivative   # should be the same as derivative

derivative: -2.33459e+10 / alt_derivative: -2.33459e+10


# Gradient Descent

Now we will write a function that performs a gradient descent. The basic premise is simple. Given a starting point we update the current weights by moving in the negative gradient direction. Recall that the gradient is the direction of *increase* and therefore the negative gradient is the direction of *decrease* and we're trying to *minimize* a cost function. 

The amount by which we move in the negative gradient *direction*  is called the 'step size'. We stop when we are 'sufficiently close' to the optimum. We define this by requiring that the magnitude (length) of the gradient vector to be smaller than a fixed 'tolerance'.

With this in mind, complete the following gradient descent function below using your derivative function above. For each step in the gradient descent we update the weight for each feature befofe computing our stopping criteria

In [11]:
from math import sqrt # recall that the magnitude/length of a vector [g[0], g[1], g[2]] is sqrt(g[0]^2 + g[1]^2 + g[2]^2)

In [12]:
def regression_gradient_descent(feature_matrix, output, initial_weights, step_size, tolerance):
    converged = False 
    weights = np.array(initial_weights) # make sure it's a numpy array
    while not converged:
        preds = predict_output(feature_matrix, weights) # compute the predictions from feature_matrix and weights using predict_output() function
        errors = preds - output                         # compute the errors as predictions - output
        gradient_sum_squares = 0.0                      # initialize the gradient sum of squares
        #
        # while we haven't reached the tolerance yet, update each feature's weight
        for ix in range(len(weights)): # loop over each weight
            # Recall that feature_matrix[:, i] is the feature column associated with weights[i]
            deriv_ix = feature_derivative(errors, feature_matrix[:, ix]) # compute the derivative for weight[i]:
            gradient_sum_squares += deriv_ix * deriv_ix # add squared value of derivative to gradient_sum_squares (for assessing convergence)
            weights[ix] -= step_size * deriv_ix         # subtract the step size times the derivative from the CURRENT weight
            
        gradient_magnitude = sqrt(gradient_sum_squares) # compute the square-root of the gradient sum of squares to get the gradient magnitude:
        if gradient_magnitude < tolerance:
            converged = True
    return weights

A few things to note before we run the gradient descent. Since the gradient is a sum over all the data points and involves a product of an error and a feature the gradient itself will be very large since the features are large (squarefeet) and the output is large (prices). So while you might expect "tolerance" to be small, small is only relative to the size of the features. 

For similar reasons the step size will be much smaller than you might expect but this is because the gradient has such large values.

# Running the Gradient Descent as Simple Regression

First let's split the data into training and test data.

In [13]:
train_data, test_data = sales.random_split(.8, seed=0)

Although the gradient descent is designed for multiple regression since the constant is now a feature we can use the gradient descent function to estimate the parameters in the simple regression on squarefeet.  
The following cell sets up the `feature_matrix`, `output`, `initial weights` and step size for the first model:

In [14]:
# let's test out the gradient descent
simple_features = ['sqft_living']
my_output = 'price'
(simple_feature_matrix, output) = get_numpy_data(train_data, simple_features, my_output)

initial_weights = np.array([-47000., 1.])
step_size = 7e-12
tolerance = 2.5e7

Next run your gradient descent with the above parameters.

In [15]:
ya_weights = regression_gradient_descent(simple_feature_matrix, output, initial_weights, step_size, tolerance)
ya_weights

array([-46999.88716555,    281.91211912])

How do your weights compare to those achieved in week 1 (don't expect them to be exactly the same)? 

Week1 values were:
  - Intercept: -47116.0766
  - Slope: 281.9588

<br />

**Quiz Question: What is the value of the weight for sqft_living -- the second element of ‘simple_weights’ (rounded to 1 decimal place)?**
  - $281.9$



Use your newly estimated weights and your `predict_output()` function to compute the predictions on all the *test* data.  
*you will need to create a numpy array of the test feature_matrix and test output first, as follows:*

In [16]:
(test_simple_feature_matrix, test_output) = get_numpy_data(test_data, simple_features, my_output)

Now compute your predictions using test_simple_feature_matrix and your weights from above.

In [17]:
ya_preds = predict_output(test_simple_feature_matrix, ya_weights)
ya_preds

array([356134.44317093, 784640.86422788, 435069.83652353, ...,
       663418.65300782, 604217.10799338, 240550.4743332 ])

**Quiz Question: What is the predicted price for the 1st house in the TEST data set for model 1 (round to nearest dollar)?**

In [18]:
pred_price_first_house = ya_preds[0] 
print(f"price: {pred_price_first_house:6.2f} / price rounded to nearest dollar: {round(pred_price_first_house, 0):6.2f}")

price: 356134.44 / price rounded to nearest dollar: 356134.00


Now that you have the predictions on test data, compute the RSS on the test data set. Save this value for comparison later. Recall that RSS is the sum of the squared errors (difference between prediction and output).

In [19]:
rss_test = np.sum(np.square(ya_preds - test_output))
rss_test

print(f"rss on test set: {rss_test} / in scientific notation: {rss_test:2.4e}")

rss on test set: 275400047593155.94 / in scientific notation: 2.7540e+14


# Running a multiple regression

Now we will use more than one actual feature. Use the following code to produce the weights for a second model with the following parameters:

In [20]:
model_features = ['sqft_living', 'sqft_living15'] # sqft_living15 is the average squarefeet for the nearest 15 neighbors. 
my_output = 'price'
(feature_matrix, output) = get_numpy_data(train_data, model_features, my_output)
initial_weights = np.array([-100000., 1., 1.])
step_size = 4e-12
tolerance = 1e9

Use the above parameters to estimate the model weights. Record these values for your quiz.

In [21]:
ya_weights_mr = regression_gradient_descent(feature_matrix, output, initial_weights, step_size, tolerance)
ya_weights_mr

array([-9.99999688e+04,  2.45072603e+02,  6.52795277e+01])

Use your newly estimated weights and the predict_output function to compute the predictions on the *test* data.  
*Don't forget to create a numpy array for these features from the test set first!*

In [22]:
(test_mr_features_matrix, test_mr_output) = get_numpy_data(test_data, model_features, my_output)

ya_preds_mr = predict_output(test_mr_features_matrix, ya_weights_mr)
ya_preds_mr

array([366651.41203656, 762662.39786164, 386312.09499712, ...,
       682087.39928241, 585579.27865729, 216559.20396617])

**Quiz Question: What is the predicted price for the 1st house in the TEST data set for model 2 (round to nearest dollar)?**

In [23]:
pred_mr_price_first_house = ya_preds_mr[0] 
print(f"price: {pred_mr_price_first_house:6.2f} / price rounded to nearest dollar: {round(pred_mr_price_first_house, 0):6.2f}")

price: 366651.41 / price rounded to nearest dollar: 366651.00


What is the actual price for the 1st house in the test data set?

In [24]:
test_data['price'][0]

310000.0

**Quiz Question: Which estimate was closer to the true price for the 1st house on the TEST data set, model 1 or model 2?**

  - [x] Model 1
  - [ ] Model 2


Now use your predictions and the output to compute the RSS for model 2 on TEST data.

In [25]:
rss_test_mr = np.sum(np.square(ya_preds_mr - test_mr_output))
rss_test_mr

print(f"rss on test set: {rss_test_mr} / in scientific notation: {rss_test_mr:2.4e}")

rss on test set: 270263446465244.06 / in scientific notation: 2.7026e+14


**Quiz Question: Which model (1 or 2) has lowest RSS on all of the test data?**

  - [ ] model 1
  - [x] model 2